### PR TITLE
[BD-24][TNL-7607] BB-3072: Simplify LTI 1.3 Launches

### PR DESCRIPTION
**Pull request by giovannicimolin**
Doing so by removing the OIDC view, which just rendered a form that performed a GET request.

**Testing instructions:**
1. Run a LTI Launch using this branch.
2. Check that the Launch is valid in the IMS tool.

**Reviewers:**
- [x] @shimulch 
- [x] @nedbat

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @giovannicimolin! I've created [BLENDED-678](https://openedx.atlassian.net/browse/BLENDED-678) to keep track of it in Jira.   More details are on the [BD-24](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1545076784/BD-24+LTI+v1.3+Improvements+Advantage+Support) project page.


When this pull request is ready, tag your edX technical lead.

File: test_settings.py [hunk 1]
```diff
@@ -11,4 +11,9 @@
...
 
 # LMS Urls - for LTI 1.3 testing
-LMS_ROOT_URL = "https://example.com"
\ No newline at end of file
+LMS_ROOT_URL = "https://example.com"
+
+# Dummy FEATURES dict
+FEATURES = {
+    'LTI_1P3_ENABLED': False,
+}
```
Comment (shimulch) [inline test_settings.py:19]: @giovannicimolin, can you add a new line at the end of this file?

File: lti_consumer/api.py [hunk 2]
```diff
@@ -53,36 +53,51 @@ def get_lti_consumer(config_id=None, block=None):
         # Since the block was passed, preload it to avoid
         # having to instance the modulestore and fetch it again.
         lti_config.block = block
+    else:
+        raise Exception('Either a config_id or block is required to get or create an LTI Configuration.')
```
Comment (shimulch) [inline lti_consumer/api.py:57]: Maybe an ``LtiError`` would be better than the ``Exception`` class?

File: lti_consumer/api.py [hunk 3]
```diff
@@ -36,12 +36,12 @@ def _get_or_create_local_lti_config(lti_version, block_location):
     return lti_config
 
 
-def get_lti_consumer(config_id=None, block=None):
+def _get_lti_config(config_id=None, block=None):
```
Comment (shimulch) [inline lti_consumer/api.py:39]: Why are we defaulting to ``None`` for ``config_id`` and ``block``? Is there any use case for that?

File: lti_consumer/api.py [hunk 4]
```diff
@@ -53,36 +53,51 @@ def get_lti_consumer(config_id=None, block=None):
...
 
 
 def get_lti_1p3_launch_info(config_id=None, block=None):
     """
     Retrieves the Client ID, Keyset URL and other urls used to configure a LTI tool.
     """
-    if not config_id and not block:
-        raise Exception("You need to pass either a config_id or a block.")
+    # Retrieve LTI Config
+    lti_config = _get_lti_config(config_id, block)
```
Comment (shimulch) [inline lti_consumer/api.py:79]: @giovannicimolin, on previous implementation we could get launch info by passing either a ``config_id`` or a ``block``. But new implementation uses ``block`` just for [caching purposes](https://github.com/edx/xblock-lti-consumer/pull/125/files#diff-6e408862179502a0ac809fab914e028b18311a17a3b0874469bd6416bc4a449fR55). This changes old behavior. Is/will there be any case that we might only have ``block`` but not ``config_id`` to get launch info?

File: lti_consumer/lti_xblock.py [hunk 5]
```diff
@@ -1357,10 +1337,20 @@ def _get_context_for_template(self):
...
-        lti_block_launch_handler = self.runtime.handler_url(self, 'lti_launch_handler').rstrip('/?')
-        if self.lti_version == 'lti_1p3':
-            lti_block_launch_handler = self.runtime.handler_url(self, 'lti_1p3_launch_handler').rstrip('/?')
+        if self.lti_version == 'lti_1p1':
+            # Set launch handler depending on LTI version
+            lti_block_launch_handler = self.runtime.handler_url(self, 'lti_launch_handler').rstrip('/?')
+        else:
+            # Runtime import since this will only run in the
+            # Open edX LMS/Studio environments.
+            from lti_consumer.api import get_lti_1p3_launch_start_url  # pylint: disable=import-outside-toplevel
```
Comment (shimulch) [inline lti_consumer/lti_xblock.py:1346]: If this import requires LMS/Studio environment, should we move this import to ``compat.py``?

**Review by shimulch — COMMENTED** [review]
@giovannicimolin, I've tested the LTI launch and it works correctly. I have some comments on the code. Could you address those?

File: lti_consumer/lti_xblock.py (hunk 5 above)
Comment (giovannicimolin) [reply lti_consumer/lti_xblock.py:1346]: @shimulch This doesn't need any LMS API, but when run inside the XBlock's test environment it'll fail because workbench is not allowed to access Django imports, so any internal `api.py` calls from the XBlock needs to be imported at runtime. 

There's a similar usage of this in this file already, here: https://github.com/edx/xblock-lti-consumer/blob/master/lti_consumer/lti_xblock.py#L838

File: lti_consumer/api.py (hunk 3 above)
Comment (giovannicimolin) [reply lti_consumer/api.py:39]: @shimulch Because we can retrieve a configuration either using the `block` or the `config_id`, and there's no need to use both.

Eventually, we'll only use `config_id` but there's no full model support for it now, hence the current implementation.
I've just added an extra method to abstract re-usable code for the deep linking implementation.

> Is there any use case for that?

Yes, LTI-AGS and other advantage APIs rely on this feature.

File: lti_consumer/api.py (hunk 4 above)
Comment (giovannicimolin) [reply lti_consumer/api.py:80]: @shimulch The behavior is exactly the same as previously, and the only thing that changed is that I have pushed the following code into a reusable function (that will be used extensively).

I've moved this code to `get_lti_config`:
```
    if config_id:
        lti_config = LtiConfiguration.objects.get(pk=config_id)
    elif block:
        lti_config = _get_or_create_local_lti_config(
            block.lti_version,
            block.location,
        )
        # Since the block was passed, preload it to avoid
        # having to instance the modulestore and fetch it again.
        lti_config.block = block
```

Comment (giovannicimolin) [reply lti_consumer/api.py:80]: @shimulch Missed your last question:

>Is/will there be any case that we might only have block but not config_id to get launch info?

Every time the API is called from whiting the XBlock, it will **only** pass the `block`, that's because there's no LTI config id storage in the LTI XBlock for now. LTI APIs only use LTI Config ID in the other hand.

We want to get rid of the Block requirement for XBlock launches, but that's on a future implementation.

**Review by shimulch — APPROVED** [review]
LGTM :+1:

- [x] I tested this on local devstack. LTI Launch works as expected.
- [x] I read through the code
- [N/A] I checked for accessibility issues
- [x] Includes documentation
- [N/A] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

**Comment by giovannicimolin** [comment]
Thanks @shimulch!

@nedbat This is ready for review. This is a part of the Deep Linking implementation from #112.

File: test_settings.py [hunk 6]
```diff
@@ -11,4 +11,9 @@
 ROOT_URLCONF = 'test_urls'
 
 # LMS Urls - for LTI 1.3 testing
-LMS_ROOT_URL = "https://example.com"
\ No newline at end of file
+LMS_ROOT_URL = "https://example.com"
+
+# Dummy FEATURES dict
+FEATURES = {
+    'LTI_1P3_ENABLED': False,
```
Comment (nedbat) [inline test_settings.py:18]: Just for my own curiosity: why did you need to add this, and why is it False?

File: README.rst [hunk 7]
```diff
@@ -298,6 +298,11 @@ Please do not report security issues in public. Send security concerns via email
 Changelog
 =========
 
+2.4.1 - 2020-12-07
```
Comment (nedbat) [inline README.rst:301]: Do we want to increment the version for each pull request?  There's no advantage to deploying these intermediate steps, right? (Other than finding out sooner if there was some bad interaction out in the real world).

**Review by nedbat — APPROVED** [review]

File: README.rst (hunk 7 above)
Comment (giovannicimolin) [reply README.rst:301]: @nedbat For this specific PR I think it's good to have a version bump.

For the LTI Deep Linking ones there's no need.

**Comment by giovannicimolin** [comment]
@nedbat this is ready for review and merge.

File: test_settings.py (hunk 6 above)
Comment (giovannicimolin) [reply test_settings.py:18]: Instead of relying on the values that the code defaults to, I've decided to explicitly add it here.
